modules/navier_stokes/src/kernels/INSMomentumBase.C

```diff
@@ -4,168 +4,183 @@
+  : INSBase(parameters),
     _component(getParam<unsigned>("component")),
-    _integrate_p_by_parts(getParam<bool>("integrate_p_by_parts")),
-
-    // Material properties
-    _mu(getMaterialProperty<Real>("mu_name")),
-    _rho(getMaterialProperty<Real>("rho_name")),
-
-    // Convective term toggling
-    _convective(getParam<bool>("convective_term"))
+    _integrate_p_by_parts(getParam<bool>("integrate_p_by_parts"))
 {
+  if (_stabilize && !_convective_term)
+    mooseError("It doesn't make sense to conduct SUPG stabilization without a convective term.");
 }
 
 Real
 INSMomentumBase::computeQpResidual()
 {
-  // The convection part, rho * (u.grad) * u_component * v.
-  // Note: _grad_u is the gradient of the _component entry of the velocity vector.
-  const Real convective_part =
-      _convective
-          ? _rho[_qp] * (_u_vel[_qp] * _grad_u[_qp](0) + _v_vel[_qp] * _grad_u[_qp](1) +
-                         _w_vel[_qp] * _grad_u[_qp](2)) *
-                _test[_i][_qp]
-          : 0.0;
-
-  // The pressure part, -p (div v) or (dp/dx_{component}) * test if not integrated by parts.
-  Real pressure_part = 0.0;
+  Real r = 0;
+
+  // viscous term
+  r += computeQpResidualViscousPart();
+
+  // pressure term
   if (_integrate_p_by_parts)
-    pressure_part = -_p[_qp] * _grad_test[_i][_qp](_component);
+    r += _grad_test[_i][_qp](_component) * weakPressureTerm();
   else
-    pressure_part = _grad_p[_qp](_component) * _test[_i][_qp];
+    r += _test[_i][_qp] * strongPressureTerm()(_component);
+
+  // body force term
+  r += _test[_i][_qp] * bodyForcesTerm()(_component);
 
-  // Call derived class to compute the viscous contribution.
-  const Real viscous_part = computeQpResidualViscousPart();
+  // convective term
+  if (_convective_term)
+    r += _test[_i][_qp] * convectiveTerm()(_component);
 
-  // Body force term
-  const Real body_force_part = -_rho[_qp] * _gravity(_component) * _test[_i][_qp];
+  if (_stabilize)
+    r += computeQpPGResidual();
 
-  return convective_part + pressure_part + viscous_part + body_force_part;
+  return r;
 }
 
 Real
-INSMomentumBase::computeQpJacobian()
+INSMomentumBase::computeQpPGResidual()
 {
   RealVectorValue U(_u_vel[_qp], _v_vel[_qp], _w_vel[_qp]);
 
-  // Convective part
-  const Real convective_part =
-      _convective
-          ? _rho[_qp] * ((U * _grad_phi[_j][_qp]) + _phi[_j][_qp] * _grad_u[_qp](_component)) *
-                _test[_i][_qp]
-          : 0.0;
+  RealVectorValue convective_term = _convective_term ? convectiveTerm() : RealVectorValue(0, 0, 0);
+  RealVectorValue viscous_term =
+      _laplace ? strongViscousTermLaplace() : strongViscousTermTraction();
+  RealVectorValue transient_term =
+      _transient_term ? timeDerivativeTerm() : RealVectorValue(0, 0, 0);
+
+  return tau() * U * _grad_test[_i][_qp] * (convective_term + viscous_term + transient_term +
+                                            strongPressureTerm() + bodyForcesTerm())(_component);
```

Comment:
Hmm... it's kind of too bad that we have to compute all those vector quantities and then throw everything except the `component`'th entry away. I guess we won't worry about it until it shows up in profiling, though...